client/rpc-spec-v2/src/chain_head/subscription.rs

```diff
@@ -0,0 +1,284 @@
+	/// of blocks pinned.
+	ExceededLimits,
+	/// Custom error.
+	Custom(String),
+}
+
+/// Inner subscription data structure.
+struct SubscriptionInner<Block: BlockT> {
+	/// The `runtime_updates` parameter flag of the subscription.
+	runtime_updates: bool,
+	/// Signals the "Stop" event.
+	tx_stop: Option<oneshot::Sender<()>>,
+	/// The blocks pinned.
+	blocks: HashSet<Block::Hash>,
+	/// The maximum number of pinned blocks allowed per subscription.
+	max_pinned_blocks: usize,
+}
+
+/// Manage the blocks of a specific subscription ID.
+#[derive(Clone)]
+pub struct SubscriptionHandle<Block: BlockT> {
+	inner: Arc<RwLock<SubscriptionInner<Block>>>,
+	/// The best reported block by this subscription.
+	/// Have this as a separate variable to easily share
+	/// the write guard with the RPC layer.
+	best_block: Arc<RwLock<Option<Block::Hash>>>,
+}
+
+impl<Block: BlockT> SubscriptionHandle<Block> {
+	/// Construct a new [`SubscriptionHandle`].
+	fn new(runtime_updates: bool, tx_stop: oneshot::Sender<()>, max_pinned_blocks: usize) -> Self {
+		SubscriptionHandle {
+			inner: Arc::new(RwLock::new(SubscriptionInner {
+				runtime_updates,
+				tx_stop: Some(tx_stop),
+				blocks: HashSet::new(),
+				max_pinned_blocks,
+			})),
+			best_block: Arc::new(RwLock::new(None)),
+		}
+	}
+
+	/// Trigger the stop event for the current subscription.
+	///
+	/// This can happen on internal failure (ie, the pruning deleted the block from memory)
+	/// or if the user exceeded the amount of available pinned blocks.
+	pub fn stop(&self) {
+		let mut inner = self.inner.write();
+
+		if let Some(tx_stop) = inner.tx_stop.take() {
+			let _ = tx_stop.send(());
+		}
+	}
+
+	/// Pin a new block for the current subscription ID.
+	///
+	/// Returns whether the value was newly inserted if the block can be pinned.
+	/// Otherwise, returns an error if the maximum number of blocks has been exceeded.
+	pub fn pin_block(&self, hash: Block::Hash) -> Result<bool, SubscriptionManagementError> {
+		let mut inner = self.inner.write();
+
+		if inner.blocks.len() == inner.max_pinned_blocks {
+			// We have reached the limit. However, the block can be already inserted.
+			if inner.blocks.contains(&hash) {
+				return Ok(false)
+			} else {
+				return Err(SubscriptionManagementError::ExceededLimits)
+			}
+		}
+
+		Ok(inner.blocks.insert(hash))
+	}
+
+	/// Unpin a new block for the current subscription ID.
+	///
+	/// Returns whether the value was present in the set.
+	pub fn unpin_block(&self, hash: &Block::Hash) -> bool {
+		let mut inner = self.inner.write();
+		inner.blocks.remove(hash)
+	}
+
+	/// Check if the block hash is present for the provided subscription ID.
+	///
+	/// Returns `true` if the set contains the block.
+	pub fn contains_block(&self, hash: &Block::Hash) -> bool {
+		let inner = self.inner.read();
+		inner.blocks.contains(hash)
+	}
+
+	/// Get the `runtime_updates` flag of this subscription.
+	pub fn runtime_updates(&self) -> bool {
```

Comment:
maybe `has_runtime_updates` is more appropriate name this function...

`runtime_updates` sounds to me like one could fetch the runtime updates which is not case here

Reply:
Yep, that better represents what the function does, thanks for the review!